python/source.py

```diff
@@ -590,12 +621,64 @@ def eig_power(self, freq):
+            )
+        elif isinstance(self.eig_band, int):
+            eig_band = self.eig_band
+
+        add_eig_src_args = [
+            self.component,
+            self.src.swigobj,
+            direction,
+            where,
+            eig_vol,
+            eig_band,
+            mp.py_v3_to_vec(
+                sim.dimensions, self.eig_kpoint, is_cylindrical=sim.is_cylindrical
+            ),
+            self.eig_match_freq,
+            self.eig_parity,
+            self.eig_resolution,
+            self.eig_tolerance,
+            self.amplitude,
+        ]
+        add_eig_src = functools.partial(
+            sim.fields.add_eigenmode_source, *add_eig_src_args
+        )
+
+        if isinstance(self.eig_band, mp.DiffractedPlanewave):
+            add_eig_src(self.amp_func, diffractedplanewave)
+        else:
+            add_eig_src(self.amp_func)
+
 
 class GaussianBeamSource(Source):
     """
     This is a subclass of `Source` and has **all of the properties** of `Source` above. However, the `component` parameter of the `Source` object is ignored. The [Gaussian beam](https://en.wikipedia.org/wiki/Gaussian_beam) is a transverse electromagnetic mode for which the source region must be a *line* (in 2d) or *plane* (in 3d). For a beam polarized in the $x$ direction with propagation along $+z$, the electric field is defined by $\\mathbf{E}(r,z)=E_0\\hat{x}\\frac{w_0}{w(z)}\\exp\\left(\\frac{-r^2}{w(z)^2}\\right)\\exp\\left(-i\\left(kz + k\\frac{r^2}{2R(z)}\\right)\\right)$ where $r$ is the radial distance from the center axis of the beam, $z$ is the axial distance from the beam's focus (or "waist"), $k=2\\pi n/\\lambda$ is the wavenumber (for a free-space wavelength $\\lambda$ and refractive index $n$ of the homogeneous, lossless medium in which the beam propagates), $E_0$ is the electric field amplitude at the origin, $w(z)$ is the radius at which the field amplitude decays by $1/e$ of its axial values, $w_0$ is the beam waist radius, and $R(z)$ is the radius of curvature of the beam's wavefront at $z$. The only independent parameters that need to be specified are $w_0$, $E_0$, $k$, and the location of the beam focus (i.e., the origin: $r=z=0$).
 
-    (In 3d, we use a ["complex point-source" method](https://doi.org/10.1364/JOSAA.16.001381) to define a source that generates an exact Gaussian-beam solution.  In 2d, we currently use the simple approximation of taking a cross-section of the 3d beam.  In both cases, the beam is most accurate near the source's center frequency.)
+    (In 3d, we use a ["complex point-source" method](https://doi.org/10.1364/JOSAA.16.001381) to define a source that generates an exact Gaussian-beam solution.  In 2d, we currently use the simple approximation of taking a cross-section of the 3d beam.  In both cases, the beam is most accurate near the source's center frequency.) To use the true solution for a 2d Gaussian Beam, use the `GaussianBeam2DSource` class instead.
```

Comment:
This makes sense to me. My only worry is will people struggle to realize the name change when they next update meep and their old scripts fail to recognize GaussianBeamSource?

Reply:
> they next update meep and their old scripts fail to recognize GaussianBeamSource

I agree. We should probably avoid any breaking changes for something like this... especially since this is preliminary. A future PR will add support for multiple frequencies etc.

Reply:
In other words, we shouldn't need to break the API at all. In the future, we can refactor and call `GaussianBeamSource` in either 2D or 3D and it just "does what it's supposed to do".

Right now that's a challenge due to the complex Hankel functions etc which aren't (readily) compatible with the current 3D codebase in C++.